### PR TITLE
Feature/identifiers

### PR DESCRIPTION
**Pull request by ppiwow-apay**
Add abstraction and swagger extensions to support the usage Strongly Typed identifiers in the code base and public API

### CODE REVIEW

File: src/Allegro.Extensions.Identifiers/Allegro.Extensions.Identifiers.AspNetCore/Swagger/StronglyTypedIdsSwaggerExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+                .Where(x => x.FullName != null && x.FullName.Contains(IdentifiersAssemblyNameSuffix))
+                .SelectMany(x => x.GetTypes())
+                .Where(x => IsAssignableToGenericType(x, typeof(IStronglyTypedId<>)) && !x.IsInterface && !x.IsAbstract)
+                .ToList();
+
+            foreach (var type in types)
+            {
+                options.MapType(type, () => new OpenApiSchema
+                {
+                    Type = "string",
```
[the remark this answers is not in the document]
Comment (rafalschmidt97) [reply src/Allegro.Extensions.Identifiers/Allegro.Extensions.Identifiers.AspNetCore/Swagger/StronglyTypedIdsSwaggerExtensions.cs:29]: `PaymentId` implements it but does it really work?

File: src/Allegro.Extensions.Identifiers/Allegro.Extensions.Identifiers.Demo/Properties/launchSettings.json [hunk 2]
```diff
@@ -0,0 +1,14 @@
+{
+  "profiles": {
+    "Allegro.Extensions.AspNetCore.Demo": {
+      "commandName": "Project",
+      "launchBrowser": true,
+      "launchUrl": "swagger",
+      "environmentVariables": {
+        "ASPNETCORE_ENVIRONMENT": "Production",
+        "Environment": "xyz"
```
Comment (rafalschmidt97) [inline src/Allegro.Extensions.Identifiers/Allegro.Extensions.Identifiers.Demo/Properties/launchSettings.json:9]: What's the point of the env here?

File: src/Allegro.Extensions.Identifiers/README.md [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+{
+    public T Value { get; }
+
+    public string ValueAsString { get; }
+}
+```
+
+To be able to use them in api or store in db without additional data structure we want still to use string in api but use strongly typed in code base:
+
+`http://serviceName/123123123/payemnt/13af7673-593c-4b7f-9d99-7c45faadb1e1`
```
Comment (rafalschmidt97) [inline src/Allegro.Extensions.Identifiers/README.md:49]: Here payment id is guid while in the /Identifiers it is int. Maybe lets not confuse devs?

**Review by rafalschmidt97 — APPROVED** [review]
LGTM

**Comment by rafalschmidt97** [comment]
Do we need to add anything to `/workflows`?

File: src/Allegro.Extensions.Identifiers/Allegro.Extensions.Identifiers.Abstractions/Allegro.Extensions.Identifiers.Abstractions.csproj [hunk 4]
```diff
@@ -0,0 +1,32 @@
+<?xml version="1.0" encoding="utf-8"?>
+<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <PackageId>Allegro.Extensions.Identifiers.Abstractions</PackageId>
+    <PackageDescription>Contains strongly typed identifiers abstractions.</PackageDescription>
+    <NoWarn>CS1574</NoWarn>
```
Comment (orgiele) [inline src/Allegro.Extensions.Identifiers/Allegro.Extensions.Identifiers.Abstractions/Allegro.Extensions.Identifiers.Abstractions.csproj:6]: An explanation for this `NoWarn` might be handy 🙏🏿

File: src/Allegro.Extensions.Identifiers/Allegro.Extensions.Identifiers.Abstractions/IStronglyTypedEntity.cs [hunk 5]
```diff
@@ -0,0 +1,12 @@
+namespace Allegro.Extensions.Identifiers.Abstractions;
+
+public interface IStronglyTypedEntity<out T, Tu>
```
Comment (orgiele) [inline src/Allegro.Extensions.Identifiers/Allegro.Extensions.Identifiers.Abstractions/IStronglyTypedEntity.cs:3]: These public interfaces have no xml docs (by contrast, `IStronglyTypedId` is documented)

File: src/Allegro.Extensions.Identifiers/Allegro.Extensions.Identifiers.Abstractions/IStronglyTypedId.cs [hunk 6]
```diff
@@ -0,0 +1,12 @@
+namespace Allegro.Extensions.Identifiers.Abstractions;
+
+/// <summary>
+/// Marker interface. Indicates if class should be treated as strongly typed identifier.
```
Comment (orgiele) [inline src/Allegro.Extensions.Identifiers/Allegro.Extensions.Identifiers.Abstractions/IStronglyTypedId.cs:4]: class -> a class
as strongly -> as a strongly

Comment (orgiele) [reply src/Allegro.Extensions.Identifiers/Allegro.Extensions.Identifiers.Abstractions/IStronglyTypedId.cs:4]: If we're mentioning a class here, should the `T` type be constrained to a class? 🤔

File: src/Allegro.Extensions.Identifiers/Allegro.Extensions.Identifiers.AspNetCore/Swagger/StronglyTypedIdsSwaggerExtensions.cs [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+using Microsoft.OpenApi.Models;
+using Swashbuckle.AspNetCore.SwaggerGen;
+
+namespace Allegro.Extensions.Identifiers.AspNetCore.Swagger;
+
+public static class StronglyTypedIdsSwaggerExtensions
+{
+    private const string IdentifiersAssemblyNameSuffix = "Identifiers";
+
+    public static IServiceCollection AddStronglyTypedIds(this IServiceCollection services)
```
Comment (orgiele) [inline src/Allegro.Extensions.Identifiers/Allegro.Extensions.Identifiers.AspNetCore/Swagger/StronglyTypedIdsSwaggerExtensions.cs:15]: No docs here - is it deliberate? Is there any PR check to enforce documenting public classes?

File: src/Allegro.Extensions.Identifiers/Allegro.Extensions.Identifiers.Demo/Controllers/TestController.cs [hunk 8]
```diff
@@ -0,0 +1,23 @@
+using System;
+using Allegro.Extensions.Identifiers.Demo.Identifiers;
+using Microsoft.AspNetCore.Mvc;
+
+namespace Allegro.Extensions.Identifiers.Demo.Controllers
+{
+    [ApiController]
+    [Route("srtrongly-typed")]
```
Comment (orgiele) [inline src/Allegro.Extensions.Identifiers/Allegro.Extensions.Identifiers.Demo/Controllers/TestController.cs:8]: typo: srtrongly -> strongly

File: src/Allegro.Extensions.Identifiers/Allegro.Extensions.Identifiers.Demo/Controllers/TestController.cs [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+    public class StronglyTypedExamples : ControllerBase
+    {
+        [HttpGet]
+        public IActionResult Get()
+        {
+            return Ok(new { UserId = UserId.Generate(), PaymentId = PaymentId.Parse("123"), OrderId = OrderId.FromGuid(Guid.NewGuid()) });
+        }
+
+        [HttpPost]
+        public IActionResult Post(UserId userId, PaymentId paymentId, OrderId orderId)
```
Comment (orgiele) [inline src/Allegro.Extensions.Identifiers/Allegro.Extensions.Identifiers.Demo/Controllers/TestController.cs:18]: Maybe we should showcase more diverse use cases here, e.g. with [FromBody], [FromRoute], [FromQuery], etc.?

File: src/Allegro.Extensions.Identifiers/Allegro.Extensions.Identifiers.AspNetCore/Swagger/StronglyTypedIdsSwaggerExtensions.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (ppiwow-apay) [reply src/Allegro.Extensions.Identifiers/Allegro.Extensions.Identifiers.AspNetCore/Swagger/StronglyTypedIdsSwaggerExtensions.cs:35]: image
not sure if this is what you asked for?

[the remark this answers is not in the document]
Comment (ppiwow-apay) [reply src/Allegro.Extensions.Identifiers/Allegro.Extensions.Identifiers.AspNetCore/Swagger/StronglyTypedIdsSwaggerExtensions.cs:35]: or you asked if for payemntId swager suggest string or int, yes? unfortunately string, will try to fix this

[the remark this answers is not in the document]
Comment (ppiwow-apay) [reply src/Allegro.Extensions.Identifiers/Allegro.Extensions.Identifiers.AspNetCore/Swagger/StronglyTypedIdsSwaggerExtensions.cs:35]: image

File: src/Allegro.Extensions.Identifiers/Allegro.Extensions.Identifiers.Abstractions/IStronglyTypedId.cs [hunk 10]
```diff
@@ -0,0 +1,12 @@
...
+/// Meziantou.Framework.StronglyTypedId that is recommended as a code generator implements this interface by default
```
[the remark this answers is not in the document]
Comment (ppiwow-apay) [reply src/Allegro.Extensions.Identifiers/Allegro.Extensions.Identifiers.Abstractions/IStronglyTypedId.cs:5]: done

File: src/Allegro.Extensions.Identifiers/Allegro.Extensions.Identifiers.Demo/Properties/launchSettings.json (hunk 2 above)
Comment (ppiwow-apay) [reply src/Allegro.Extensions.Identifiers/Allegro.Extensions.Identifiers.Demo/Properties/launchSettings.json:9]: copy paste, fixed

File: src/Allegro.Extensions.Identifiers/README.md (hunk 3 above)
Comment (ppiwow-apay) [reply src/Allegro.Extensions.Identifiers/README.md:49]: switched to order sample fully

File: src/Allegro.Extensions.Identifiers/Allegro.Extensions.Identifiers.Abstractions/IStronglyTypedId.cs (hunk 6 above)
Comment (ppiwow-apay) [reply src/Allegro.Extensions.Identifiers/Allegro.Extensions.Identifiers.Abstractions/IStronglyTypedId.cs:4]: T is just base primitive type like Guid, string, int maybe, not sure how to constraint it. changed word to type

File: src/Allegro.Extensions.Identifiers/Allegro.Extensions.Identifiers.AspNetCore/Allegro.Extensions.Identifiers.AspNetCore.csproj [hunk 11]
```diff
@@ -0,0 +1,35 @@
+<?xml version="1.0" encoding="utf-8"?>
```
[the remark this answers is not in the document]
Comment (ppiwow-apay) [reply src/Allegro.Extensions.Identifiers/Allegro.Extensions.Identifiers.AspNetCore/Allegro.Extensions.Identifiers.AspNetCore.csproj:1]: thinking about different types of documentation (api) generator than swagger maybe?

File: src/Allegro.Extensions.Identifiers/Allegro.Extensions.Identifiers.Demo/Controllers/TestController.cs (hunk 9 above)
Comment (ppiwow-apay) [reply src/Allegro.Extensions.Identifiers/Allegro.Extensions.Identifiers.Demo/Controllers/TestController.cs:18]: done

File: src/Allegro.Extensions.Identifiers/Allegro.Extensions.Identifiers.Abstractions/Allegro.Extensions.Identifiers.Abstractions.csproj (hunk 4 above)
Comment (ppiwow-apay) [reply src/Allegro.Extensions.Identifiers/Allegro.Extensions.Identifiers.Abstractions/Allegro.Extensions.Identifiers.Abstractions.csproj:6]: here or in code? unfortunately, Meziaontou generates summaries with some strange cref to base class that is invalid ... this warning disables this.

Comment (orgiele) [reply src/Allegro.Extensions.Identifiers/Allegro.Extensions.Identifiers.Abstractions/Allegro.Extensions.Identifiers.Abstractions.csproj:6]: as discussed - we'd add a comment here in csproj

**Review by orgiele — APPROVED** [review]

File: src/Allegro.Extensions.Identifiers/Allegro.Extensions.Identifiers.AspNetCore/Allegro.Extensions.Identifiers.AspNetCore.csproj (hunk 11 above)
[the remark this answers is not in the document]
Comment (ppiwow-apay) [reply src/Allegro.Extensions.Identifiers/Allegro.Extensions.Identifiers.AspNetCore/Allegro.Extensions.Identifiers.AspNetCore.csproj:1]: i'll leave it for further purposes - still related to asp.net as swagger is rather used in this kind of projects

File: src/Allegro.Extensions.Identifiers/Allegro.Extensions.Identifiers.Abstractions/IStronglyTypedEntity.cs (hunk 5 above)
Comment (rafalschmidt97) [reply src/Allegro.Extensions.Identifiers/Allegro.Extensions.Identifiers.Abstractions/IStronglyTypedEntity.cs:6]: Good idea Szymek

**Pull request merged**